### ISSUE TITLE
OBS Support

### ISSUE THREAD

**Issue by braewoods** [ungoogled-software/ungoogled-chromium-archlinux#70]
I have been setting up the Debian derived packaging up for automated OBS builds. While doing so I noticed that Arch is a possible option for it so I was wondering if there's any value in that for our Arch users. Thoughts?

Edit: Now that I think about it I'm not sure automated builds are of much value here as Arch tends to promote building your own packages. Just thought I would throw that in here.

**Comment by wchen342** [comment]
Can OBS send back building status to Github as a CI? If so then we can use it here too otherwise I have set up a running Actions already which I think is better for CI purposes because it is integrated with pull requests.

**Comment by braewoods** [comment]
TLDR: It can probably be done with some ingenuity since the OBS API allows log retrieval but it would be limited to wherever you can safely trigger a build. And it would be limited to one build per project per package so even if it would work for pull requests it wouldn't scale too well.

I don't think it has any form of CI integration. Even if it did the only way I found to setup the automated builds was in response to direct pushes and tag publishing since the OBS API requires authentication secrets. These are not available to pull requests apparently as they could be leaked by a malicious change. If anyone cares to research it in greater depth then here is the API documentation: https://build.opensuse.org/apidocs/index

I did setup the Debian system to send normal pushes to a separate project for "development builds" whereas published tags go to the normal project. But I didn't include any kind of build monitoring because, although it should be possible via OBS API calls, I didn't see any value in it for the Debian system since I can just check it from the website. I only use it for publishing official builds or doing test builds before I publish a tag.

**Comment by Eloston** [comment]
I think we can have both:

* Actions is very integrated with GitHub. While it has good deployment features too, it's not great for building large software like Chromium (or maybe [the GitHub Workflow](https://github.com/ungoogled-software/ungoogled-chromium-archlinux/blob/master/.github/workflows/build.yml) is not as scary/complicated as it looks). However, Actions seems pretty good for CI, which would be nice to have when automated testing comes around. Before that, I'm not sure what we could use Actions for other than to initiate OBS builds.
* OBS seems pretty good at receiving package source code, building them, and publishing them on their service. As @braewoods explains, I don't think OBS is made for CI use cases.

**Comment by braewoods** [comment]
Ok, I just thought of something else. It is possible to scale the system as I set it up if the user creates their own OBS account, sets up the secrets for it, and doesn't mind it using their home project. This allows them to use it within the context of their fork. If this ever catches on I could make the script more flexible to allow a different project prefix to be used.

**Comment by wchen342** [comment]
So what I was thinking is the cost of maintaining two very different systems. I agree that for publishing releases OBS shall be better and it won't bee too hard to hook it with tag push or something. Also we still need to solve the account problem (even users can build using their own account, if we want to release some kind of "official build" we need to manage an organization account of some kind).

BTW how long does OBS take to build chromium? I wonder whether it is much faster.

**Comment by jstkdng** [comment]
It is possible to trigger rebuilds with github actions, I already set it up on my fork. Took some time to get used to the api but it is working really good now.
https://github.com/jstkdng/ungoogled-chromium-archlinux/actions/runs/90633843
You could make another action that monitors the build status and check if it finished. Regarding the secrets, you only need a basic http auth header to trigger a rebuild.

**Comment by jstkdng** [comment]
@wchen342 obs used to take half an hour when jumbo was a thing, now it takes ~2.5 hours

**Comment by wchen342** [comment]
@jstkdng That's way faster than Actions. I will look into how I can turn it into some kind of CI because the building time of Actions is really a pain.

**Comment by jstkdng** [comment]
@wchen342 This projects structure would have to change in order to work with OBS's limitations though, otherwise you would have to maintain 2 PKGBUILDS.

**Comment by braewoods** [comment]
@wchen342 It varies. I've had builds finish in about 90 minutes. Others take up to 6 hours. It depends on your luck regarding what builder takes your job.

If we did go for this I was expecting to integrate it ultimately into the OBS account I made just for our use. If the need arises I can look into granting access to other user accounts to the projects or packages they require. I'm not sure how fine-grained OBS is but it does have lists of users so I get the feeling it's possible.

**Comment by braewoods** [comment]
For reference, here's the files I made for OBS:

https://github.com/ungoogled-software/ungoogled-chromium-debian/blob/debian_sid/.github/workflows/obs-upload.yml
https://github.com/ungoogled-software/ungoogled-chromium-debian/blob/debian_sid/.github/workflows/obs-upload.sh

**Comment by wchen342** [comment]
@braewoods That script helps me understand how it works. For a minute I though the building process cannot be customized. If it can then most problem can be solved.
But I am still not confused with the account thing. My understanding is for the auto-build of *this repository* it shouldn't use any personal account because otherwise it essentially linked the project to that person, so we need an account of `ungoogled-software` which will handle all the build (including Debian and Arch versions). Am I missing anything there?

**Comment by braewoods** [comment]
@wchen342 No you're not; it's probably confusion over why I mentioned the other accounts. You need the login credentials of an OBS account to upload any build requests and to perform most API calls. However there is a token API you can use as well but it seemed so limited as to be worthless so I didn't look into it much. See here: https://build.opensuse.org/apidocs/index#23

We don't need to have people create a personal account. We can reuse the same credentials from the Debian setup by adding them to the secrets for this repository. I mentioned it because we may have other reasons to give some access to personal accounts; e.g., maintenance operations or attempting to diagnose build failures.

The account I setup for this is named for this is called ungoogled_chromium. The Debian repository has the location of the two projects currently in use by it.

**Comment by braewoods** [comment]
Oh and regarding build.script? I think that's Debian specific so if you need to customize stuff for Arch it will probably have to go into the PKGBUILD itself. I don't think this will be a problem though.

**Comment by wchen342** [comment]
@braewoods Ah I see. It makes sense now. I will start looking into how to build the Arch version (or if you have time you can write a template too).

**Comment by braewoods** [comment]
Someone already appears to have OBS builds. Perhaps they can be cross-referenced with the Debian stuff.

https://build.opensuse.org/package/show/home:justkidding:arch/ungoogled-chromium-ozone
https://build.opensuse.org/package/show/home:justkidding:arch/ungoogled-chromium

**Comment by jstkdng** [comment]
@braewoods that's me btw
you could fork the package and add it to your ungoogled-chromium user and edit stuff there.

**Comment by braewoods** [comment]
@jstkdng Quite possibly but I'm not really that involved with Arch stuff so I mostly just wanted to bring it up for discussion here. Though I do think a system like mine that generates the necessary files for the OBS side is probably the best option for long term maintenance. I only see it being useful to hand edit them when there's an issue that needs investigating.

**Comment by jstkdng** [comment]
@braewoods I agree with you on that. My approach is simpler though since there is no 100 line script required.
I only have 2 files on the OBS, the rest is generated by the obs itself. It will pull all the files from github whenever I push to my fork, that way I dont have to care about the OBS anymore.

**Comment by braewoods** [comment]
I was looking through the PKGBUILD and got the feeling these uses of curl to retrieve information from the repositories have an issue. They pull from the master branch instead of the submodule so you may have issues from using it instead of the actual tag used in the repository. An approach that pulls directly from the submodule or even just uses templated values would be preferable.

**Comment by braewoods** [comment]
@wchen342 Also worth noting that OBS has no known upper time limit on how long a build job can take. It should be good for up to 24 hours if it was needed.

**Comment by braewoods** [comment]
@jstkdng Any ideas why my prototype Arch builds keep failing with makepkg trying to run pacman with sudo? I thought all the dependencies were already installed and I can't see why it's behaving like this when it doesn't in my Arch containers. See here: https://build.opensuse.org/build/home:ungoogled_chromium:testing/Arch/x86_64/ungoogled-chromium-arch/_log

**Comment by jstkdng** [comment]
the obs doesnt source the pkgbuild, it parses the depends and optdepends and manually installs them. You have to manually add all dependencies for it to build.

**Comment by braewoods** [comment]
So I have to throw in all the implicit dependencies as well?

**Comment by braewoods** [comment]
So I am going to see if I can transition to using an independent repository to handle all the OBS logic. This is likely the most practical option for controlling access to the credentials and only exposing them to what needs it.

**Comment by wchen342** [comment]
I think we had the discussion before but I forgot the results so I'm going to asking again. So OBS does not allow proprietary ffmpeg codes, then do you guys use the "cleaned" version of chromium sources on OBS? Or you just use the normal version?

**Comment by jstkdng** [comment]
I use the same sources from opensuse's chromium package so I should be clear on that end.

**Comment by wchen342** [comment]
What numbered do you use in `_constraints`? I put 16GB physical memory but still faced random aborts which I think come from insufficient memories. The strange thing is it sometimes finished the build successfully without a problem under the same constraint.

**Comment by jstkdng** [comment]
I'm using 16gb of ram and 16gb of storage and that is enough for building. I'm not sure if there are workers with more ram. Maybe you could change something on your build process (?
https://github.com/jstkdng/aur/blob/56b3f3afd0f4b633370a89f1cc18041114b38b90/.github/workflows/obs_files#L13

**Comment by wchen342** [comment]
Turns out I need 24+GB memory. I think it is an old version of GCC eating more memory than clang.

**Comment by jstkdng** [comment]
huh, are you trying to use obs for android builds?

**Comment by wchen342** [comment]
No I was building for Fedora. They use GCC in upstream so I followed.

**Comment by braewoods** [comment]
It's implemented now for this repository. Closing.
